HPCC-31353 Report the slowest 5 activies in the roxie complete line

ghalliday opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

ghalliday committed:
HPCC-31353 Add override attributes

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
HPCC-31353 Report the slowest 5 activies in the roxie complete line

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

mckellyln commented on roxie/ccd/ccdlistener.cpp:
```diff
@@ -945,6 +945,61 @@ extern void updateAffinity(unsigned __int64 affinity)
+            ids.append(slowActivityIds[i]);
+            formatStatistic(times, cycle_to_nanosec(slowActivityTimes[i]), SMeasureTimeNs);
+        }
+        s.appendf(", slowActivities={ ids=[%s] times=[%s] }", ids.str(), times.str());
```
Not "slowActivities" but perhaps "slowestActivities" ?
What if they are all about the same ?  Is there more we can provide ?  As in:
complete in 11861 msecs ...
slowActivities={ ids=[45,195,75,63,15] times=[10.196s,10.059s,10.035s,10.028s,10.021s] }

richardkchapman commented:
My main concern is what is the performance impact of gathering this information. mergeStats may be called a lot, especially in a child query scenario. Is the information going to be useful enough, often enough? 

Not sure I would recommend allowing configuration of the number, as picking a higher number will negatively impact the performance more

ghalliday commented:
Conclusions from discussion:
- It would be better to only do this if the query was above a certain threshold/SLA
- Even better would be to generate a stats workunit for a query which exceeded the SLA - with a limit of only doing it once every minute/5 minutes.

ghalliday commented:
@richardkchapman I have added some timing tests.  It has an impact of ~1ns for every activity that isn't in the top 5 and about 10ns for each activity that is.  So a impact of ~50us for a very complex query.  That is almost certainly lower than the impact of aggregating the rest of the stats.  It is only called when the final results are aggregated, so child queries etc. will not impact it.  
After further reflection I think it is worthwhile because many roxie queries are not soapcall bound, and this provides some useful debugging information when there is a problem - full stats are much better when rerunning the query.

Thoughts/opinions?
@mckellyln renamed slow to slowest, rebased and squashed.

mckellyln reviewed (approved):
ok, looks good.

ghalliday commented:
@mckellyln would this be better to only report if the slowest activity was above a certain threshold (e.g., 10ms), or is it always useful?

mckellyln commented:
@ghalliday yes - I think good idea to skip this if slowest activity was less than some configurable threshold (10 ms default).

ghalliday committed:
Add a minimum time to the slowest roxie activies

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday commented:
Added an extra guard condition, ignoring all activities < 10ms.  (Compare ignoring case.)

mckellyln commented on roxie/ccd/ccdmain.cpp:
```diff
@@ -1320,6 +1321,9 @@ int CCD_API roxie_main(int argc, const char *argv[], const char * defaultYaml)
         updateAffinity(affinity);
 
+        unsigned __int64 minimumInterestingActivityMs = topology->getPropInt64("@minimumInterestingActivityMs", 10);
+        minimumInterestingActivityCycles = nanosec_to_cycle(minimumInterestingActivityMs * 1'000'000);
```
Nice use of c++14 separator :-)

mckellyln reviewed (approved):
Approved
